neil

```diff
@@ -270,15 +270,17 @@
         edn-nodes (edn-nodes edn-string)
         lib (:lib opts)
         lib (symbol lib)
-        git? (or (:sha opts)
-                 (:latest-sha opts))
+        explicit-git? (or (:sha opts)
+                          (:latest-sha opts))
+        [version git?] (if explicit-git?
+                         [(or (:sha opts)
+                              (latest-github-sha lib)) true]
+                         (cond
+                           (:version opts) [(:version opts) false]
+                           (latest-clojars-version lib) [(latest-clojars-version lib) false]
+                           (latest-mvn-version lib) [(latest-mvn-version lib) false]
+                           (latest-github-sha lib) [(latest-github-sha lib) true]))
```

Comment:
Not happy with this -- hoping we can find something cleaner.

1. stuff like `(latest-clojars-version lib)` is repeated twice
2. we're also downloading things twice.

Reply:
Suggestion from Slack - consider core.async for this.

Reply:
What is the actual problem that you would need core.async for this?

Reply:
I don't think we need core.async for anything here. This is just normal single-threaded logic. You can just use a `let` to bind a value. You don't have to do everything in one cond.

Reply:
We could save some time if we executed all three requests in parallel. But perhaps that doesn't make too much of a difference.

Reply:
I would say clojars -> mvn -> git sequentially would be good start. Let's save the optimizations (except double requests) for later. Double requests can be prevented by just introducing more nesting and conditionals.

Reply:
solution with local variables pushed now. Was this what you had in mind? Seems to work.

```clojure
[version git?] (if explicit-git?
                 [(or (:sha opts)
                      (latest-github-sha lib)) true]
                 (or
                  (when-let [v (:version opts)]
                    [v false])
                  (when-let [v (latest-clojars-version lib)]
                    [v false])
                  (when-let [v (latest-mvn-version lib)]
                    [v false])
                  (when-let [v (latest-github-sha lib)]
                    [v true])))

```

Reply:
Yup